apps/emqx_rule_engine/src/emqx_rule_engine.erl

```diff
@@ -185,29 +181,32 @@ update_rule(Params = #{id := RuleId}) when is_binary(RuleId) ->
     case get_rule(RuleId) of
         not_found ->
             {error, not_found};
-        {ok, #{created_at := CreatedAt}} ->
-            parse_and_insert(Params, CreatedAt)
+        {ok, ?META_DATA(MetaData)} ->
+            parse_and_load_rule(Params, MetaData#{updated_at => now_ms()})
     end.
 
 -spec delete_rule(RuleId :: rule_id()) -> ok.
 delete_rule(RuleId) when is_binary(RuleId) ->
     gen_server:call(?RULE_ENGINE, {delete_rule, RuleId}, ?T_CALL).
 
--spec insert_rule(Rule :: rule()) -> ok.
-insert_rule(Rule) ->
-    gen_server:call(?RULE_ENGINE, {insert_rule, Rule}, ?T_CALL).
+-spec load_rule_hook_and_metrics(Rule :: rule()) -> ok.
+load_rule_hook_and_metrics(Rule) ->
+    gen_server:call(?RULE_ENGINE, {load_rule, Rule}, ?T_CALL).
 
 %%----------------------------------------------------------------------------------------
 %% Rule Management
 %%----------------------------------------------------------------------------------------
 
 -spec get_rules() -> [rule()].
 get_rules() ->
-    get_all_records(?RULE_TAB).
+    [Rule || {_Id, Rule} <- maps:to_list(get_rules_map())].
```

Comment:
Q: The docs say that getting a value from persistent_terms doesn't trigger a copy, but I'm not sure if we're converting it to a list via `maps:to_list/1` if it triggers a copy of all the rules?

> Term lookup (using [get/1](https://www.erlang.org/doc/man/persistent_term.html#get-1)) is done in constant time and without taking any locks, and the term is not copied to the heap (as is the case with terms stored in ETS tables).

Reply:
Yes, but less than before. My understanding that here the two lists will be constructed (one temporary, due to `maps:to_list/1`, second as a result of list comprehension), both having the length equal to the total number of rules. Yet, list cells will still reference persistent term heap (as they are not being modified).